### ISSUE TITLE
Issues with using the Android back button when nesting a DrawerNavigator in a StackNavigator

### ISSUE THREAD

**Issue by Elethier** [react-navigation/react-navigation#3941]
### Current Behavior

I am working on a project with a team using `react-navigation`, and we're using a `DrawerNavigator` in conjunction with a `StackNavigator` to add a drawer menu to our project. The project is targeting both Android and iOS. We've noticed an issue when attempting to navigate to another page in our stack navigator from a menu selection on the drawer menu. Here is a GIF, on Android, demonstrating the behavior.

![react navigation bug android](https://user-images.githubusercontent.com/2686895/38519792-959772e8-3c06-11e8-8a24-01bd47e15d85.gif)

After selecting a text input on a page that was opened from the drawer menu, if I use the back button to return back to the home page, the drawer is closed instead of opened, and there are a few more screens on the navigation stack than I would expected.

### Expected Behavior

Here is a GIF of the same application running on iOS, which demonstrates the expected behavior:

![react navigation bug ios](https://user-images.githubusercontent.com/2686895/38520017-1e9ab528-3c07-11e8-8519-5a21d801100a.gif)

After selecting a text input a page that was opened from the drawer menu, pressing the back navigation buttons in the header returns the user to the home page of the app with the drawer open. There is no additional navigation items on the stack.

### How to reproduce

I have reproduced the issue in a [sample project](https://github.com/Elethier/react-navigation-drawer-bug). To summarize what I'm doing in the sample project, here's my navigation configuration:

```
const DrawerStack = DrawerNavigator({
  'HomePage': { screen: HomePage },
}, {
  contentComponent: DrawerMenu
})

const AppNavigator = StackNavigator({
  'Home': { screen: DrawerStack },
  'MenuPage': { screen: MenuPage },
  'InnerMenuPage': { screen: InnerMenuPage }
})
```

To open the drawer from a component, I'm calling `this.props.navigation.navigate('DrawerOpen')`. To go to the menu page from a component, I'm calling `this.props.navigation.navigate('MenuPage')`. To navigate back in the stack, I'm calling `this.props.navigation.goBack()`.

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | 1.5.11
| react-native     | 0.54.0
| node             | v9.11.1
| npm      | 5.8.0

### Investigation Notes
I've spent a good amount of time investigating the issue, so I thought I'd provide some notes.

After selecting the text input on Android, the text input gains focus. When the text input loses focus, Android moves the focus to the next focusable item in the view hierarchy. In this case, it's the native `DrawerLayout` created by the `DrawerNavigator`. Since the drawer layout has focus, it handles the back button press instead of React Native, and this causes a `DrawerClose` navigation action to occur, instead of a `Navigation/BACK` action. This mangles the stack and results in the extra items on the stack. 

I wasn't able to find a good way to fix the focus behavior, so I ended up implementing a workaround: I removed the use of `react-native-drawer-layout-polyfill` and used `react-native-drawer-layout` on both platforms instead. This avoids the issue by using a Javascript drawer layout instead of a native Android one. I'm not sure that workaround is a good long-term solution.

**Comment by Elethier** [comment]
Thanks for the reply @brentvatne. I checked out the Snack and I also updated the sample project to use version 2.0.0 of react-navigation. While the specific issue I mentioned has been fixed, I'm seeing some new issues with the drawer.

- I went to the inner menu page and started type. Then I hit back to close the keyboard. Then I hit back again to close the inner menu page. I then found it takes two back button presses instead of one back button press to close the menu page.
- After going to the inner menu page, typing something in, and then mashing the back button to get back to the home page of the app, I can't open the drawer anymore.

Should I open some separate issues for these problems?
